// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// The following only applies to changes made to this file as part of YugaByte development.
//
// Portions Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.

syntax = "proto3";

package yb.cdc;

option java_package = "org.yb.cdc";

import "yb/common/common.proto";
import "yb/common/common_net.proto";

message ProducerTabletListPB {
  // List of tablet ids, start and end keys for a given producer table.
  repeated string tablets = 1;
  repeated bytes start_key = 2;
  repeated bytes end_key = 3;
}

message ProducerSchemaPB {
  // The last Producer schema version that was EquivalentForDataCopy with this Consumer.
  uint32 validated_schema_version = 1;
  // The upcoming Producer schema for ingest.  If present, Consumer needs ALTER for compatibility.
  uint32 pending_schema_version = 2;
  SchemaPB pending_schema = 3;
}

message StreamEntryPB {
  // Map from consumer to producer tablet id.
  map<string, ProducerTabletListPB> consumer_producer_tablet_map = 1;
  string consumer_table_id = 2;
  string producer_table_id = 3;
  bool local_tserver_optimized = 4;
  ProducerSchemaPB producer_schema = 5;
}

message ProducerEntryPB {
  // Map from stream id to metadata for that stream.
  map<string, StreamEntryPB> stream_map = 1;
  repeated HostPortPB master_addrs = 2;
  repeated HostPortPB DEPRECATED_tserver_addrs = 3; // See issue #11989.
  bool disable_stream = 4; // [default = false] implicit in proto3
}

message ConsumerRegistryPB {
  // Map from producer universe uuid to metadata for that producer.
  map<string, ProducerEntryPB> producer_map = 1;
  // Enable the replication of the txn status table for ordered and atomic multi shard transaction
  // support.
  bool enable_replicate_transaction_status_table = 2;
}
